File: analyze-ocr.js
const fs = require('fs');
const path = require('path');
const { createWriteStream } = require('fs');

const recipePath = 'c:\\Users\\dajoh\\Documents\\code\\Medland-Cooks.github.io\\medland-cooks\\resources\\print-recipies';
const outputFile = 'c:\\Users\\dajoh\\Documents\\code\\Medland-Cooks.github.io\\ocr-raw-scores.csv';

function getOCRScore(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    let score = 100;
    const issues = [];

    // 1. Character corruption (high-byte characters)
    let garbledCount = 0;
    for (let i = 0; i < content.length; i++) {
      if (content.charCodeAt(i) > 127) {
        garbledCount++;
      }
    }
    if (garbledCount > 5) {
      const penalty = Math.min(garbledCount / 10, 15);
      score -= penalty;
      issues.push('Corrupted');
    }

    // 2. Fractions and units
    const fractionPattern = /\b\d+\s?\/\s?\d+\b/g;
    const fractions = (content.match(fractionPattern) || []).length;

    const unitPattern = /\b(cup|cups|tbsp|tablespoon|tsp|teaspoon|oz|ounce|lb|pound|ml|gram|g)\b/gi;
    const units = (content.match(unitPattern) || []).length;

    if (units === 0) {
      score -= 10;
      issues.push('NoUnits');
    } else if (fractions === 0) {
      score -= 3;
      issues.push('NoFractions');
    } else if (fractions > 3) {
      score += 3;
      issues.push('GoodFractions');
    }

    // 3. Punctuation
    const punctPattern = /[.,;:]/g;
    const punctTotal = (content.match(punctPattern) || []).length;
    const lines = content.split('\n');
    const lineCount = lines.length;

    if (lineCount > 5) {
      const punctPerLine = punctTotal / lineCount;
      if (punctPerLine < 0.1) {
        score -= 5;
        issues.push('LowPunctuation');
      }
    }

    // 4. Common cooking terms
    const cookingTerms = ['butter', 'sugar', 'flour', 'egg', 'salt', 'pepper', 'milk', 'cream', 'oil', 'bake', 'mix', 'combine'];
    let foundTerms = 0;
    for (const term of cookingTerms) {
      if (new RegExp(`\\b${term}\\b`, 'i').test(content)) {
        foundTerms++;
      }
    }

    if (foundTerms < 3) {
      score -= 5;
      issues.push('FewTerms');
    }

    score = Math.max(0, Math.min(100, score));

    return {
      score: Math.round(score * 10) / 10,
      issues: issues.join('|'),
      fractions,
      units,
      lines: lineCount,
      punctuation: punctTotal
    };
  } catch (e) {
    return null;
  }
}

function getGrade(score) {
  if (score >= 90) return 'A';
  if (score >= 80) return 'B';
  if (score >= 70) return 'C';
  if (score >= 60) return 'D';
  return 'F';
}

// Main execution
console.log('Reading files...');
const files = fs.readdirSync(recipePath).filter(f => 
  fs.statSync(path.join(recipePath, f)).isFile()
).sort();

const fileCount = files.length;
console.log(`Processing ${fileCount} OCR recipe files...`);

const results = [];
let processed = 0;

for (const filename of files) {
  processed++;
  if (processed % 100 === 0) {
    console.log(`  Progress: ${processed} / ${fileCount}`);
  }

  const filePath = path.join(recipePath, filename);
  const scoreData = getOCRScore(filePath);

  if (scoreData) {
    const grade = getGrade(scoreData.score);
    results.push({
      Filename: filename,
      Score: scoreData.score,
      Grade: grade,
      Issues: scoreData.issues,
      Fractions: scoreData.fractions,
      Units: scoreData.units,
      Lines: scoreData.lines,
      Punctuation: scoreData.punctuation
    });
  }
}

// Export to CSV
const csvContent = [
  ['Filename', 'Score', 'Grade', 'Issues', 'Fractions', 'Units', 'Lines', 'Punctuation'],
  ...results.map(r => [r.Filename, r.Score, r.Grade, r.Issues, r.Fractions, r.Units, r.Lines, r.Punctuation])
]
  .map(row => row.map(cell => `"${String(cell).replace(/"/g, '""')}"`).join(','))
  .join('\n');

fs.writeFileSync(outputFile, csvContent);
console.log(`\nSaved to: ${outputFile}`);

// Statistics
const gradeCounts = { A: 0, B: 0, C: 0, D: 0, F: 0 };
for (const r of results) {
  gradeCounts[r.Grade]++;
}

console.log('\nGRADE DISTRIBUTION:');
for (const grade of ['A', 'B', 'C', 'D', 'F']) {
  const count = gradeCounts[grade];
  const pct = ((count / fileCount) * 100).toFixed(1);
  console.log(`  ${grade}: ${count} files (${pct}%)`);
}

// Top A-grade files
const aFiles = results
  .filter(r => r.Grade === 'A')
  .sort((a, b) => b.Score - a.Score)
  .slice(0, 20);

console.log(`\nTOP A-GRADE FILES (${aFiles.length} found):`);
if (aFiles.length > 0) {
  aFiles.slice(0, 10).forEach(r => {
    console.log(`  ${r.Filename}: Score ${r.Score}`);
  });
} else {
  console.log('  No A-grade files found');
}

// Top B-grade files
const bFiles = results
  .filter(r => r.Grade === 'B')
  .sort((a, b) => b.Score - a.Score)
  .slice(0, 10);

console.log(`\nSAMPLE B-GRADE FILES (${bFiles.length} found):`);
if (bFiles.length > 0) {
  bFiles.slice(0, 5).forEach(r => {
    console.log(`  ${r.Filename}: Score ${r.Score}`);
  });
}

console.log('\nDone!');

File: analyze-ocr.py
#!/usr/bin/env python3
import os
import re
import csv

recipe_path = r"c:\Users\dajoh\Documents\code\Medland-Cooks.github.io\medland-cooks\resources\print-recipies"
output_file = r"c:\Users\dajoh\Documents\code\Medland-Cooks.github.io\ocr-raw-scores.csv"

def get_ocr_score(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
            content = f.read()
    except:
        return None
    
    score = 100
    issues = []
    
    # 1. Check for corrupted characters (high-byte encodings)
    garbled_count = sum(1 for c in content if ord(c) > 127)
    if garbled_count > 5:
        penalty = min(garbled_count / 10, 15)
        score -= penalty
        issues.append("Corrupted")
    
    # 2. Fractions and units
    fraction_pattern = r'\b\d+\s?/\s?\d+\b'
    fractions = len(re.findall(fraction_pattern, content))
    
    unit_pattern = r'\b(cup|cups|tbsp|tablespoon|tsp|teaspoon|oz|ounce|lb|pound|ml|gram|g)\b'
    units = len(re.findall(unit_pattern, content, re.IGNORECASE))
    
    if units == 0:
        score -= 10
        issues.append("NoUnits")
    elif fractions == 0:
        score -= 3
        issues.append("NoFractions")
    elif fractions > 3:
        score += 3
        issues.append("GoodFractions")
    
    # 3. Punctuation
    punct_count = len(re.findall(r'[.,:]', content))
    lines = content.split('\n')
    line_count = len(lines)
    
    if line_count > 5:
        punct_per_line = punct_count / line_count
        if punct_per_line < 0.1:
            score -= 5
            issues.append("LowPunctuation")
    
    # 4. Common cooking terms
    cooking_terms = ['butter', 'sugar', 'flour', 'egg', 'salt', 'pepper', 'milk', 'cream', 'oil', 'bake', 'mix', 'combine']
    found_terms = sum(1 for term in cooking_terms if re.search(fr'\b{term}\b', content, re.IGNORECASE))
    
    if found_terms < 3:
        score -= 5
        issues.append("FewTerms")
    
    score = max(0, min(100, score))
    
    return {
        'score': score,
        'issues': '|'.join(issues),
        'fractions': fractions,
        'units': units,
        'lines': line_count,
        'punctuation': punct_count
    }

def get_grade(score):
    if score >= 90:
        return 'A'
    elif score >= 80:
        return 'B'
    elif score >= 70:
        return 'C'
    elif score >= 60:
        return 'D'
    else:
        return 'F'

# Get all files
files = sorted([f for f in os.listdir(recipe_path) if os.path.isfile(os.path.join(recipe_path, f))])
file_count = len(files)
print(f"Processing {file_count} OCR recipe files...")

results = []
for idx, filename in enumerate(files, 1):
    if idx % 100 == 0:
        print(f"  Progress: {idx} / {file_count}")
    
    file_path = os.path.join(recipe_path, filename)
    score_data = get_ocr_score(file_path)
    
    if score_data:
        grade = get_grade(score_data['score'])
        results.append({
            'Filename': filename,
            'Score': score_data['score'],
            'Grade': grade,
            'Issues': score_data['issues'],
            'Fractions': score_data['fractions'],
            'Units': score_data['units'],
            'Lines': score_data['lines'],
            'Punctuation': score_data['punctuation']
        })

# Export to CSV
with open(output_file, 'w', newline='') as f:
    writer = csv.DictWriter(f, fieldnames=['Filename', 'Score', 'Grade', 'Issues', 'Fractions', 'Units', 'Lines', 'Punctuation'])
    writer.writeheader()
    writer.writerows(results)

print(f"Saved to: {output_file}")

# Statistics
grade_counts = {'A': 0, 'B': 0, 'C': 0, 'D': 0, 'F': 0}
for r in results:
    grade_counts[r['Grade']] += 1

print("\nGRADE DISTRIBUTION:")
for grade in ['A', 'B', 'C', 'D', 'F']:
    count = grade_counts[grade]
    pct = (count / file_count) * 100 if file_count > 0 else 0
    print(f"  {grade}: {count} files ({pct:.1f}%)")

# Top A-grade files
a_files = sorted([r for r in results if r['Grade'] == 'A'], key=lambda x: x['Score'], reverse=True)[:20]
print(f"\nTOP A-GRADE FILES (showing up to 20):")
if a_files:
    for r in a_files[:10]:
        print(f"  {r['Filename']}: Score {r['Score']}")
else:
    print("  No A-grade files found")

print("\nDone!")
